src/transformers/trainer_tf.py

```diff
@@ -0,0 +1,509 @@
+        else:
+            raise ValueError(
+                "Allowed optimizers are adam, adamw, adadelta, rmsprop or sgd. Given: {}".format(self.optimizer_name)
+            )
+
+    def _create_checkpoint_manager(self, checkpoint_path, max_to_keep=5, load_model=True):
+        """
+        Create a checkpoint manager in order to be able to make the training
+        fault-tolerant.
+        Args:
+          max_to_keep: the maximum number of checkpoints to keep in the
+            checkpoint path.
+          load_model: if we want to start the training from the latest checkpoint.
+        """
+        ckpt = tf.train.Checkpoint(optimizer=self.optimizer, model=self.model)
+        self.model.ckpt_manager = tf.train.CheckpointManager(ckpt, checkpoint_path, max_to_keep=max_to_keep)
+
+        if load_model:
+            ckpt.restore(self.model.ckpt_manager.latest_checkpoint)
+
+    def _evaluate_during_training(self):
+        """
+        Evaluate the model during the training at the end of each epoch.
+        """
+        num_batches = 0
+        test_step = 1
+
+        for batch in self.datasets["validation"]:
+            test_step = tf.convert_to_tensor(test_step, dtype=tf.int64)
+            self._distributed_test_step(batch)
+            num_batches += 1
+
+            with self.test_writer.as_default():
+                tf.summary.scalar("loss", self.test_loss_metric.result(), step=test_step)
+
+            if test_step % self.validation_steps == 0:
+                break
+
+            test_step += 1
+
+    def train(self):
+        """
+        Train method to train the model.
+        """
+        with self.strategy.scope():
+            tf.summary.trace_on(graph=True, profiler=True)
+
+            step = 1
+            train_loss = 0.0
+
+            for epoch in range(1, self.epochs + 1):
+                total_loss = 0.0
+                num_batches = 0
+
+                for batch in self.datasets["train"]:
+                    step = tf.convert_to_tensor(step, dtype=tf.int64)
+                    total_loss += self._distributed_train_step(batch)
+                    num_batches += 1
+
+                    with self.train_writer.as_default():
+                        tf.summary.scalar("loss", total_loss / num_batches, step=step)
+
+                    if step == 1:
+                        with self.train_writer.as_default():
+                            tf.summary.trace_export(name="training", step=step, profiler_outdir=self.log_path)
+
+                    if step % 10 == 0:
+                        logger.info("Epoch {} Step {} Loss {:.4f} Train Accuracy {:.4f}".format(epoch, step.numpy(), total_loss.numpy() / num_batches, self.train_acc_metric.result()))
+
+                    if step % 100 == 0:
+                        ckpt_save_path = self.model.ckpt_manager.save()
+                        logger.info("Saving checkpoint for step {} at {}".format(step, ckpt_save_path))
```

Comment:
The frequency for logging and checkpointing should be controllable

Reply:
Yes, it is planned as well!